### ISSUE TITLE
Temporarily Set Configuration For Save Operation

### ISSUE THREAD

**Issue by jedmao** [Microsoft/vscode#65663]
I'm maintaining the [EditorConfig](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig) extension and am in dire need of a feature that would eliminate some issues with respect to the following settings and potentially more in the future (e.g., end of line setting):

- `files.insertFinalNewline`
- `files.trimFinalNewlines`
- `files.trimTrailingWhitespace`

Specifically, I need a way to short-circuit / override these settings temporarily for a file-save operation, because allowing vscode to perform any kind of trimming is destructive and either irreversible or extremely hacky to fix. It's like asking me to unpunch someone.

Basically, I need a way to change the way that vscode does a built-in save operation on a per-file basis (which is how EditorConfig works). Pre-save TextEdits are not good enough, because whatever I do before a save could potentially be wiped out with vscode's destructive trimming after the fact.

Another option, of course, would be to make EditorConfig a built-in feature of vscode, which I think makes all the sense in the world.

**Comment by segevfiner** [comment]
VS Code already has [TextEditorOptions](https://code.visualstudio.com/api/references/vscode-api#TextEditorOptions) which holds similar settings to be overridden by an extension.

I guess it can be extended with tri-state properties (`null` - use config, `true` - run, ignoring config, `false` - don't run, ignoring config) that allow the extension to override whether these built-in pre save actions will be ran.

This will also allow the extension to reuse the built-in actions rather than reimplementing them in it's own code.

**Comment by Tyriar** [comment]
@bpasero any insights into the real solution would be appreciated, I was convinced this was a bug in VS Code itself until I confirmed --disable-extensions worked and searched for the culprit extension. Editor config is a very popular extension (even featured right now) and this bug is forcing me to uninstall.

**Comment by bpasero** [comment]
Adding @jrieken to this thread who added the API for doing something pre-save. 

> Another option, of course, would be to make EditorConfig a built-in feature of vscode, which I think makes all the sense in the world.

Yes, probably a good idea at some point.

**Comment by Tyriar** [comment]
There's also this discussion happening on the extension repo: https://github.com/editorconfig/editorconfig-vscode/issues/208#issuecomment-451728131

**Comment by Tyriar** [comment]
I think this is the underlying cause of the issue https://github.com/Microsoft/vscode/issues/66214

**Comment by Tyriar** [comment]
@jedmao is this still relevant given https://github.com/Microsoft/vscode/issues/66214?

**Comment by segevfiner** [comment]
This issue is about not being able to override per TextEditor whether the pre-save operations run. Which is what the editorconfig extension needs in order to be implemented correctly. Otherwise if the pre-save operations are configured to run via the settings, they will run even if the editorconfig file says otherwise.

@Tyriar what you mentioned is a different issue.

**Comment by jedmao** [comment]
@segevfiner is correct. This issue is absolutely still relevant.

**Comment by jedmao** [comment]
Any updates on this guy?

**Comment by segevfiner** [comment]
Looks like the VS Code project itself is now recommending the EditorConfig extension. Would be nice if it was prioritized adding/changing what's needed so it can be implemented fully and accurately.

**Comment by nrayburn-tech** [comment]
@jrieken, sharing my thoughts on a solution for this.  Please correct me if I am wrong anywhere, or you disagree.

My understanding of the problem is that an extension may have a configuration to enforce some coding style that directly conflicts with the vscode save participant settings.
For example, eslint has `eol-last`, editorconfig has `insert_final_new_line`, and vscode has `files.insertFinalNewLine` all of which can conflict in one way or another.  In the case of eslint and editorconfig, these setting values can be different for each file in a folder, whereas the `files.insertFinalNewLine` is applied to the entire workspace.
Eslint modifications are run using `editor.codeActionsOnSave.source.fixAll`, editorconfig modifications are done with `TextDocumentWillSaveEvent.waitUntil` with the transformations.  In both of these cases, the `insertFinalNewLine` save participant will be the final change applied.

Solutions
1. Original solution, allow an extension to disable any save participants by id.  This gives the extension too much control over the save participants.  This isn't an acceptable solution.

2. Solution 2, modify the save participant values from just `true`/`false` to be `true`/`false`/`true with extension overrides`/`false with extension overrides` (possibly specify extensions by id that can override, so that not all extensions have access.) This would work for both the modification methods that eslint and editorconfig use.

3. Solution 3, modify the order of the save participants so that the `editor.codeActionsOnSave` are always run last, so that the extensions have the final say about transformations.  One downside to this is it only works for extensions using `codeActionsOnSave`, it wouldn't work for the method that editorconfig uses.  Of course extensions could be updated to use whatever is necessary.

In each of the above solutions, the extension would need a way to pass the information in per model.

Does solution 2 or 3 work for you?  Your thoughts on a different solution?

**Comment by nrayburn-tech** [comment]
> editor-config could know about some conflicting defaults in vscode

This is simple enough for the current save participants.

> VS Code could try to explain what's happening before save and why

This is probably more complex than checking for conflicts and I am not sure that this is needed in the case of editorconfig.

With the current save participants that VS Code has, there probably aren't any changes that need to be made in VS Code.  Extensions can update and notify the user of a conflict through notifications, status bar, logs, etc.  These notifications can guide the user to configuring their workspace settings to avoid the conflicting configurations.  Is this right?

**Comment by SamB** [comment]
Well, I don't think anybody is really asking VS Code to worry about what happens when eslint config and .editorconfig disagree; just that there needs to be a way that either extension can override the default or user-level VS Code settings. (Overriding workspace-specific settings wouldn't really hurt either, but it's less essential as the user could reasonably be instructed "don't do that, then" w.r.t conflicting workspace settings, just as for conflicts between editorconfig and jslint.)

**Comment by Anutrix** [comment]
The EditorConfig bug almost made me permanently get rid of VSCode because there was no way to find which extension was doing 'Remove space on save'.

Wish there was some logs to find out which extension is messing up when I do something basic like save a file.

**Comment by sanmai-NL** [comment]
You can save _without formatting_ too. The meaning of saving is clearly documented.
